[clang] Constexpr for __builtin_shufflevector and __builtin_convertvector

Destroyerrrocket opened the pull request:
Summary:

This patch adds constexpr support for __builtin_shufflevector and __builtin_convertvector.

A small oddity encountered was that the arg to the intrinsics may be an lvalue without any sort of implicit cast of any kind. I solved this through the EvaluateVectorOrLValue function, which treats the lvalue as if it was in an rvalue cast, which gets me the desired vector.

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,138 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  if (!Result.isVector()) {
+    return false;
+  }
+  return true;
```
Just use `return Result.isVector()` ?

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,138 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  // but in constexpr we need to choose a value. We'll choose 0.
+  if (index >= TotalElementsInAVector * 2) {
+    index = 0;
+  }
```
Is there any chance we can retain the undef? Could we insert an undef element?

Destroyerrrocket replied:
If there is, I unfortunately don't know how...

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,138 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  APSInt IndexVal;
+  if (!EvaluateInteger(IndexExpr, IndexVal, Info)) {
+    return false;
+  }
```
(style) unnecessary braces

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,138 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  VectorType const *DestVecTy = E->getType()->castAs<VectorType>();
+  if (!DestVecTy) {
+    return false;
+  }
```
(style) unnecessary braces

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,138 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+
+  SmallVector<APValue, 4> ResultElements;
+  ResultElements.reserve(Source.getVectorLength());
+  for (unsigned EltNum = 0; EltNum < Source.getVectorLength(); ++EltNum) {
```
Pull out repeated Source.getVectorLength() calls?

RKSimon commented on clang/test/Sema/constant-builtins-2.c:
```diff
@@ -302,3 +302,64 @@ extern __typeof__(__builtin_expect(0, 0)) bi0;
+vector4int vectorShuffle6 = __builtin_shufflevector(
+    vector8intConst, vector8intConst, 0, 2, 4, 6);
+vector8int vectorShuffle7 =
+    __builtin_shufflevector(vector4intConst1, vector4intConst2, 0, 2, 4, 6, 1, 3, 5, 7);
```
What could we do to check the result values of the constant expression? Could we bitcast a char4 to a uint value and static assert the result for instance?

Destroyerrrocket replied:
You're right! we can indeed make something like:
constexpr vector4char v = {1,2,3,4};
static_assert(std::bit_cast<unsigned>(v) == 0x4030201, "equal");
I had initially dismissed this as the operator[] is not implemented for vectors. This is endian dependent, is there a particular way I should handle this?

RKSimon replied:
Maybe use the same 'LITTLE_END' define approach used in clang\test\SemaCXX\constexpr-builtin-bit-cast.cpp?

How much work will it to be to eventually implement the vector operator[] as a constexpr as well do you think?

Destroyerrrocket replied:
Yep, I'll copy that over.

Well, in that case I would have to care about lvalues (in my code I just needed the values to return a new copy, so I could ignore all that machinery for now). I'd have to learn how these work more in depth. I don't think they'd be necessarily super complicated, and that would make for an ideal follow-up for me! (looks like this would be part of: https://github.com/llvm/llvm-project/issues/41806, and also this also looks interesting: https://github.com/llvm/llvm-project/issues/30794)

I think this is a pretty interesting set of tasks to work on!

Destroyerrrocket replied:
But just for this validation, I'd definitiely take the bitcast approach

RKSimon reviewed (commented):
A few minors but a frontend specialist really needs to review this

RKSimon commented:
Please can you rename constat_builtins_vector.cpp -> constant_builtins_vector.cpp

Destroyerrrocket commented:
(sorry about the typo!)

erichkeane commented on clang/test/Sema/constant_builtins_vector.cpp:
```diff
@@ -0,0 +1,99 @@
+// RUN: %clang_cc1 -verify -std=c++2a -fsyntax-only %s
+// expected-no-diagnostics
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Can we clean up this test?  All of the macros doing the 'check' generation make this pretty unreadable.

Destroyerrrocket replied:
I just really didn't want to type out all the combinations. I can expand the macros out, np!

Destroyerrrocket replied:
I'm sorry @sethp, I don't follow your explanation, I'm not familiar enough with the testing infrastructure in Clang. If you think that will make the code look cleaner while still not having to spell out all convinations, I'm all for it!

Destroyerrrocket replied:
If you could provide a toy example, I can do that for this test

sethp replied:
My suggestion was not so much a way to avoid spelling them all out as to trick the preprocessor into doing (a lot of) the work. I'll try to put together a toy example in the next couple of days, but if you run `clang -E` on the existing file and commit the resulting spellings you'll get much of the way to what I was thinking. 

Would that address your concern, @erichkeane?

sethp replied:
Well, I got a quick demo of what I was thinking put together. For your consideration: https://github.com/llvm/llvm-project/commit/17c78a28b4f441bf284bcf6935061938ec908f31

It did & didn't work out quite the way I'd hoped: I'd forgotten just how, uh, _peculiar_ the preprocessor is, especially if you're trying to get it to generate human-readable output. There might be a way to do it nicely (adding `clang-format` into the pipeline?), or it might not be an idea worth pursuing at all. I'll leave the final determination up to you & @erichkeane .

erichkeane replied:
No, i don't think that is worth doing.  I'm still disturbed by how difficult it is to figure out exactly what/how we are testing things, but perhaps just some detailed comments on the macros are sufficient.

erichkeane commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
   return Success(APValue(ResultElements.data(), ResultElements.size()), E);
 }
 
+static bool EvaluateVectorOrLValue(APValue &Result, EvalInfo &Info,
```
This interface to this function seems really odd for the name?  Is the name inaccurate, or is this using 'false' to mean 'failure' AND 'not a vector'?  

It seems to me that this needs to be something like "EvaluateLValueVector" or something?

Destroyerrrocket replied:
> is the name inaccurate, or is this using 'false' to mean 'failure' AND 'not a vector'?
Indeed, false is being treated as a failure and not a vector.
> It seems to me that this needs to be something like "EvaluateLValueVector" or something?
I'm not entirely sure if the way I implemented the handling of an lvalue as a parameter is even the correct way to go with, so it might very well be that the reason it sticks out is because I'm doing something wrong. I guess I'll have to wait for someone who has dealt with lvalues in constexpr to tell me where I went wrong

Destroyerrrocket replied:
I think that these errors should be covered by convertvector.c, I'll add one of the missing diagnostics you pointed out. One already exists

Destroyerrrocket replied:
Went with the EvaluateAsRValue

erichkeane reviewed (commented):
Hopefully someone with more knowledge of the const evaluator (@shafik and @tbaederr have experience lately!) can poke their head in, but a few things from my look over.

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2702,7 +2702,8 @@ static bool checkFloatingPointResult(EvalInfo &Info, const Expr *E,
                                    APFloat &Result) {
-  assert(isa<CastExpr>(E) || isa<CompoundAssignOperator>(E));
+  assert(isa<CastExpr>(E) || isa<CompoundAssignOperator>(E) ||
+         isa<ConvertVectorExpr>(E));
```
(style) assertion message

sethp commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  if (!EvaluateInteger(IndexExpr, IndexVal, Info))
+    return false;
+
+  uint32_t index = IndexVal.getZExtValue();
+  // The spec says that -1 should be treated as undef for optimizations,
+  // but in constexpr we need to choose a value. We'll choose 0.
+  if (index >= TotalElementsInAVector * 2)
+    index = 0;
```
I don't think this quite implements `__builtin_shufflevector`; consider:

```c++
typedef char vector4char __attribute__((__vector_size__(4)));
typedef char vector8char __attribute__((__vector_size__(8)));

__builtin_shufflevector((vector4char){}, (vector8char){}, 10);
```

That should produce a single-element `char` vector whose element has value `0`: https://godbolt.org/z/9655x1oc6 (though it looks like there's another issue elsewhere in Sema that's not currently accepting that).

Also, the `>=` suggests to me that it'll also produce values for out-of-bounds indexes. Happily, this already aborts (with a useful diagnostic):

```c++
__builtin_shufflevector((vector8char){}, (vector8char){}, 20);
```

So, for that part it's probably enough here to clarify with an `assert` that `index` is less the sum of the vectors' lengths (or the -1 sentinel).

sethp replied:
Also, I think just picking a value here isn't quite right. Consider:

```c++
constexpr auto f() {
    return __builtin_shufflevector((vector8char){7}, (vector8char){}, -1);
}
```

This implementation will always produce a value of `7` for `f()[0]` when called in a constant context, right? But it's some flavor of U{nspecified,ndefined}B, so the value produced at runtime is unknown. As I understand it the evaluator is required to bail out on UB, to prevent exactly that kind of mismatch: it's preferable for the interpreter to say "I can't do anything with this" than to do something that would only match some of the time.

So, as I understand it, there's two options here:

1. `return false` with some diagnostic saying "no `-1` indexes in a constant context", or
2. use `APValue::Indeterminate` (or `APValue::None`?) for this slot in the vector (oh, I guess _that's_ why there's an APValue per element).

The latter would allow an expression like:

```c++
static_assert(__builtin_shufflevector((vector8char){7}, (vector8char){}, 0, -1)[0] == 7);
```

but not

```c++
// should fail to produce a constant value
static_assert(__builtin_shufflevector((vector8char){7}, (vector8char){}, 0, -1)[1]); 
```

(interestingly, neither `gcc` nor `clang` works completely as I'd expect here; clang doesn't yet support constant-time `[..]` vector element accesses, and `gcc` actually allows the second construction to produce `0`, which it probably shouldn't).

Destroyerrrocket replied:
<source>:6:12: error: first two arguments to '__builtin_shufflevector' must have the same type
    6 |     return __builtin_shufflevector((vector4char){}, (vector8char){}, 10)[0];
      |            ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

Latest trunk seems to also think that both input vectors should match in size. I'll add an llvm_unreachable for the case where it is out of bounds

Destroyerrrocket replied:
APValue::Indeterminate seems to be exactly what I was looking for! I'll read up on this, thank you. I assumed that such a thing did not exist, but I'm glad to be proven wrong

Destroyerrrocket replied:
Operator[] needs lvalue support for vectors, I think I commented with @RKSimon that it would be a really useful thing for validations. For now, I'm static casting to get arround this

Destroyerrrocket replied:
Bad news... seems like APValue::Indeterminate and None are both intentionally disallowed. I'll go with the second option then, make the user choose what element they want to use as a source in constexpr. They can always `if consteval`

sethp replied:
re `Operator[]` there's already a PR for doing that in constant contexts, I think: https://github.com/llvm/llvm-project/pull/72607 . Definitely agree it'd be helpful for validations! 

re: `APValue::Indeterminate`, I think you're talking about how top-level indeterminate values get rejected by clang as not being a constant expression, right? Dang, that both makes sense and feels unfortunate: thanks for giving it a try, at least!

Destroyerrrocket replied:
Great, let's hope operator[] makes it.

The error comes from somewhere in the depths of ExprConst, it intentionally tells you that undefined values are indeed illegal in constexpr code. I agree with you, a shame it makes sense!

sethp reviewed (commented):
Hi @Destroyerrrocket, thanks for your work getting vectors into constexpr-land! I'm also curious about that, so I did my best to give you a helpful review. 

The only part I'd be uncomfortable about merging is the handling of `-1` sentinels / indeterminate values: everything else I expect you to take as no more than a gentle suggestion.

Destroyerrrocket commented:
@sethp This is an amazing review! I won't be able to work on this for 2 weeks, but thank you a lot for your work!

RKSimon commented:
@Destroyerrrocket reverse-ping - are you still working on this?

Destroyerrrocket commented:
@RKSimon Hi, Sorry I have not reached out! Yes, I plan to do some work on this in the next two weeks; on a normal work week I end up way too tired to work more in the weekend, if you need this and want to take over just let me know, I don't want to stall anyone!

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  const FPOptions FPO = E->getFPFeaturesInEffect(Info.Ctx.getLangOpts());
+
+  auto SourceLen = Source.getVectorLength();
+  SmallVector<APValue, 4> ResultElements;
```
I chose 4 as plenty of times I've had to deal with SIMD code it has been either SSE code dealing with floats and doubles and AVX code mainly dealing with doubles, but this certainly is a biased worldview; not all applications deal with doubles and certainly from time to time I've had to code for avx512.

In any case, I think that doubling the smallvector would not be ideal stacksize-wise, and a normal vector with a reserve call would be more than enough to cover all cases. We always know the size and a normal allocation should be fine.

APValue is indeed expensive, but it is also convenient; I think I'm going to stick with it for this PR, and if very long vectors ever get supported this might need to be improved upon. Should I leave a comment explaining that to the reader?

Destroyerrrocket replied:
Looking at how other code does it, it seems like I also based my decision on the fact that all other vector code uses a size of 4. I think that just for consistency, it is ok to leave it as it is.

sethp replied:
Yeah, I don't have much of an idea how to evaluate the tradeoffs here: I was hoping you might, and that you'd teach me :)

I agree that it's out of scope for this PR, and that sticking with the 4 for consistency's sake is as good a number as any!

Destroyerrrocket replied:
304 bytes is definitely inside what I'd be slightly worried to allocate without thinking into the stack, but having looked at it I think it's going to be fine. We're not the only case of smallVectors with APValues, and I'm sure that if it ever becomes a problem the default stack size can be modified (or the code can be refactored, I'm sure that if that ever happens it's not going to be "just" the fault of a smallvector of APValues). I thank you for pointing it out, as I simply did not check!

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+                                  DestTy, Result.getInt());
+    }
+  }
+  return false;
```
I certainly don't know about any such case...

Destroyerrrocket replied:
it seems like the only one missing is _BitInt(8) or bigger. I'll just add support for that

sethp replied:
It looks like the bit-precise support "just worked"? At least, I see a test for it, but not any changes in this function. That's neat!

I'm still leaning towards adding a diagnostic here, though: I'm imagining a future standard or compiler version adding a new vector-capable type (_Complex, maybe? One of the fixed-point types?), and callers of `__builtin_convertvector(...)` with an that type will just see some flavor of "must be initialized by a constant expression," despite  the docs' promise that "This builtin can be used within constant expressions." 

I'm not sure how to test it—maybe using an incomplete type?—but I'm also pretty comfortable with an un-exercised `Info.FFDiag(...)` call here too. That said, this certainly does work with everything I can think to throw at it today, though, so I'm open to calling it a future follow-up idea too. What do you think?

Destroyerrrocket replied:
I agree that adding a diagnostic here, even if unreachable, is more than just desired, I'm well convinced on the future-proofing argument. I will not add a test because I simply don't have a way to reach this, I believe.

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+                                QualType ElemType, APValue const &VecVal1,
+                                APValue const &VecVal2, unsigned EltNum,
+                                APValue &Result) {
+  unsigned const TotalElementsInAVector = VecVal1.getVectorLength();
```
Aren't both vectors the same size? This code assumes that based on the spec. If They are not the same size, I need to revisit this, but otherwise I think the name is just fine. Maybe I'll rename it to `TotalElementsInInputVector` so it is a bit more descriptive

sethp replied:
I didn't think the vectors had to be the same length: both clang and gcc document the extension to require that the vectors have "a compatible element type" ([gcc](https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html)) or "are vectors that have the same element type" ([clang](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-shufflevector)). 

However, while gcc does accept a shuffle of e.g. an 8-element and 4-element vector (for a 12-element "input"), clang seems to reject it: https://godbolt.org/z/3ooszMKbc

So maybe the thing to do here is add an `assert(VecVal1.getVectorLength() == VecVal2.getVectorLength())`? And/or, change the documentation to make it clear it's "same element type _and length_"?

An alternative to consider would be to plumb both the `constexpr` and codegen pieces to accept a wider range of vectors for arguments 1 and 2, but I'm assuming the current lowering is just using LLVM's `shufflevector`, which itself only supports homogeneous arguments. To me, that smells like something best addressed as its own unit of work and far outside the scope of this PR, but I wanted to include it for completeness.

Destroyerrrocket replied:
I see! It seems like Gcc has extra characteristics then. given that at some point it could happen that such a thing needs to be supported, let's actually rewrite the code to support both sizes. Beware, I won't be able to test it, as Sema will prevent it from reaching constexpr.

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+                                   APValue &Result) {
+  if (SourceTy->isIntegerType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = APValue(APFloat(0.0));
+      return HandleIntToFloatCast(Info, E, FPO, SourceTy, Original.getInt(),
+                                  DestTy, Result.getFloat());
```
I think that beyond some differences in memory layout, I'd expect the performance to be about the same for the normal case. The bad case might be able to save a bit of performance by avoiding a copy (happening inside the APValue) (that probably will get optimized anyway). Without measuring, I think this is performant enough, and I value code being short and easy enough to read (not that the proposed code is bad in any way; If I saw it in a PR I wouldn't even blink)

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+}
+
+bool VectorExprEvaluator::VisitShuffleVectorExpr(const ShuffleVectorExpr *E) {
+  APValue VecVal1;
+  const Expr *Vec1 = E->getExpr(0);
+  if (!EvaluateVectorOrLValue(VecVal1, Info, Vec1, Vec1->getType()))
+    return false;
+  APValue VecVal2;
+  const Expr *Vec2 = E->getExpr(1);
+  if (!EvaluateVectorOrLValue(VecVal2, Info, Vec2, Vec2->getType()))
+    return false;
```
I 100% Agree! I'll look if concatenating the vectors can be done in an efficient manner!

Destroyerrrocket replied:
I could not find a way that would not require copying (the move operator is just a copy in this case) about 600 bytes of data. I think I'll just leave it as is, as just making a new APValue to insert values to does not seem to solve much

sethp replied:
Ah, too bad: thanks for looking into it and reporting back!

Destroyerrrocket commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+
+  SmallVector<APValue, 4> ResultElements;
+  ResultElements.reserve(TotalElementsInOutputVector);
+  for (unsigned EltNum = 0; EltNum < TotalElementsInOutputVector; ++EltNum) {
```
getShuffleMaskIdx is exactly what I should be using! Thanks for pointing that out

Destroyerrrocket commented:
Whenever you have a change, I think this is ready for a second pass. Sorry for the delay! ^-^ If I missed something, do let me know

Destroyerrrocket commented:
I'll rebase when I have more time, around next week (I know that ideally you'd prefer to see it already rebased)

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10318,6 +10318,8 @@ def err_shufflevector_nonconstant_argument : Error<
   "index for __builtin_shufflevector must be less than the total number "
   "of vector elements">;
+def err_shufflevector_minus_one_is_undefined_behavior_constexpr : Error<
+  "index for __builtin_shufflevector must be within the bounds of the input vectors in a constexpr context. An index of -1 at position %0 was found. -1 is only allowed at runtime.">;
```
```suggestion
  "index for __builtin_shufflevector not within the bounds of the input vectors; index of -1 found at position %0 not permitted in a constexpr context">;
```

Leave the 'only allowed at runtime' for the documentation (and document it?).

erichkeane reviewed (commented):
I would love it if @sethp  can do another review, but only 1 nit from me.

sethp commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -10961,6 +10968,118 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  if (!EvaluateAsRValue(Info, Vec2, VecVal2))
+    return false;
+
+  VectorType const *DestVecTy = E->getType()->castAs<VectorType>();
+  if (!DestVecTy)
+    return false;
+
+  QualType DestElTy = DestVecTy->getElementType();
```
Oh, I missed this before: `castAs` isn't ever supposed to produce a null pointer; I think it's roughly the llvm not-RTTI spelling of `static_cast` (if I'm using any of those words right), whereas `dyn_cast` (which `Type` wraps as `getAs`) is more akin to a fail-able `dynamic_cast`.

So this ought to be equivalent to:

```suggestion
  VectorType const *DestVecTy = E->getType()->castAs<VectorType>();
  QualType DestElTy = DestVecTy->getElementType();
```

But I'll leave it up to you whether that's a change worth making here. Mostly this is me trying to flex my "do I remember what the difference between `Type::getAs` and `Type::castAs` is" muscle, and it's offered to you in case you're interested in the same exercise :smile:

Destroyerrrocket replied:
That's good to know indeed :)

sethp reviewed (commented):
The __builtin_shufflevector impl LGTM (one tiny note below), though I still owe @Destroyerrrocket a worked example on the test side.

I have one suggestion for `__builtin_convertvector`, but overall that LGTM too: I think we're really close to merging this!

RKSimon commented:
Other than fixing the ReleaseNotes.rst conflict is there anything outstanding on this now?

Destroyerrrocket commented:
There are some minor comments I believe, I just haven't had any time to
look at them. I'm sorry about that! Hopefully I'll be able to give it the
nudge it needs this month, but looking at the calendar I'm not sure I'd
count on it (no vacations coming)

El lun, 15 abr 2024, 17:14, Simon Pilgrim ***@***.***>
escribió:

> Other than fixing the ReleaseNotes.rst conflict is there anything
> outstanding on this now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/76615#issuecomment-2057104839>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGBMPSGR2YKS5P552RKENULY5PVFTAVCNFSM6AAAAABBHQSW6GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJXGEYDIOBTHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sethp commented:
That sounds right to me; the tasks I see are:

- [ ] Expanding the test cases for `__builtin_convertvector`
- [ ] Rewording the diagnostic for -1 in `__builtin_shufflevector`
- [ ] Rewording the comment for -1 in `__builtin_shufflevector`
- [ ] A stylistic change to `__builtin_convertvector`
- [ ] Considering adding a diagnostic for `__builtin_convertvector` in case we "fall through" to the bottom (which we think is currently impossible)
- [ ] ReleaseNotes.rst

Of those, I think the only one that isn't pretty much purely mechanical is whether or not we add the extra diagnostic. 

Let me know if I missed anything, or if there's anything I can do to help get this one over the finish line: thanks so much @Destroyerrrocket for all your work already!

Destroyerrrocket commented:
Addressed the comments; I don't really like the expanded macro code, it generates 600 lines of really repetitive code, so I left the macros as a source of truth that I'd use to generate the cases in the future.

sethp reviewed (approved):
LGTM!

Barring any objections, I'll plan to hit the squash & merge button on this one come Monday (29 April) morning. 

Thank you @Destroyerrrocket for your thoroughness and patience! I enjoyed learning my way through the details with you :smile:

Destroyerrrocket commented:
It's been a pleasure! Expect to see me again taking the operator[] accessor
(if no one else takes it) by June aprox

El mié, 24 abr 2024, 20:26, sethp ***@***.***> escribió:

> ***@***.**** approved this pull request.
>
> LGTM!
>
> Barring any objections, I'll plan to hit the squash & merge button on this
> one come Monday (29 April) morning.
>
> Thank you @Destroyerrrocket <https://github.com/Destroyerrrocket> for
> your thoroughness and patience! I enjoyed learning my way through the
> details with you 😄
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/76615#pullrequestreview-2020629069>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGBMPSEDG56GCEBWA63BF4LY672MXAVCNFSM6AAAAABBHQSW6GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMRQGYZDSMBWHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

philnik777 commented:
I just wanted to say thanks for working on this. I'd love to enable the vectorization branches in libc++ during constant evaluation. It's always a great way to have confidence in the correctness of the code, and this brings us a step closer to having fewer `__libcpp_is_constant_evaluated()` in the code base.

Destroyerrrocket committed:
Mark PR#76615 as integrated

Destroyerrrocket commented:
Done!

kamaub commented:
This changeset causes failures on the [clang-ppc64be-linux-multistage](https://lab.llvm.org/buildbot/#/builders/93) bot with build [19869](https://lab.llvm.org/buildbot/#/builders/93/builds/19869) and [clang-ppc64be-linux-test-suite](https://lab.llvm.org/buildbot/#/builders/231) bot with build [22932](https://lab.llvm.org/buildbot/#/builders/231/builds/22932). It noticeably passes [clang-ppc64le-linux-multistage](https://lab.llvm.org/buildbot/#/builders/121) on build [41191](https://lab.llvm.org/buildbot/#/builders/121/builds/41191), they are big and little endian versions of each other. 
Please investigate and supply a fix or revert your patch as soon as possible, thank you.

Please let me know if I can assist with this. This appears to be the correct PR for https://github.com/llvm/llvm-project/commit/2903df02fb3c057849aaa796a91289b01950a5f0 and so I have posted there as well.

Destroyerrrocket commented:
(I think) I know what the issue is; The test has a macro to generate the little/big endian constants. But during review we expanded them because that helped readability. The generating macros are still there, I understand that this is kinda urgent so if someone has 10 minutes to remove the expanded macro code and uncomment the macro usages, the tests should pass again in big endian mode. 
Beware that at the bottom there are a few extra manual tests. If no one can right now, I think it's easier if we revert and merge again at our own pace :)

sethp commented:
Ah, that sounds right: I'll tweak the tests and make sure the test case runs on both big & little-endian systems.

Destroyerrrocket commented:
Thank you @sethp ! I've marked the lines that need changing in the commit

sethp commented:
Oh, I did it slightly differently: I just backed out the "last step" of the macro expansion by hand (`1 ?` turns out not to be a pattern that shows up too much on purpose).  

That's in https://github.com/llvm/llvm-project/commit/347a02b408567ba15fdc68646129c1f5de97ab7e ; thanks @kamaub !

Destroyerrrocket commented:
Great, works for me!
